peps/pep-0667.rst

```diff
@@ -40,72 +62,239 @@ The current implementation of ``locals()`` and ``frame.f_locals`` is slow,
+    ...
+    >>> f()
+    1
+    1
+    1
+
+This situation was more common in Python 3.10 and earlier
+versions, as merely installing a tracing function was enough
+to trigger implicit calls to ``PyFrame_LocalsToFast()`` after
+every line of Python code. However, it can still happen in Python
+3.11+ depending on exactly which tracing functions are active
+(e.g. interactive debuggers intentionally do this so that changes
+made at the debugging prompt are visible when code execution
+resumes).
+
+All of the above comments in relation to ``exec()`` apply to 
+*any* attempt to mutate the result of ``locals()`` in optimized
+scopes, and are the main reason that the ``locals()`` builtin
+docs contain this caveat:
+
+    Note: The contents of this dictionary should not be modified;
+    changes may not affect the values of local and free variables
+    used by the interpreter. 
+
+Two options were considered to replace this confusing behaviour:
+
+* make ``locals()`` return write-through proxy instances (similar
+  to ``frame.f_locals``)
+* make ``locals()`` return genuinely independent snapshots so that
+  attempts to change the values of local variables via ``exec()``
+  would be *consistently* ignored without any of the caveats
+  noted above.
+
+The PEP chooses the second option for the following reasons:
+
+* returning independent snapshots in optimized scopes preserves
+  the Python 3.0 change to ``exec()`` that resulted in attempts
+  to mutate local variables via ``exec()`` being ignored in most
+  cases
+* the distinction between "``locals()`` gives an instantaneous
+  snapshot of the local variables in optimized scopes, and
+  read/write access in other scopes" and "``frame.f_locals``
+  gives read/write access to the local variables in all scopes,
+  including optimized scopes" allows the intent of a piece of
+  code to be clearer than it would be if both APIs granted
+  full read/write access in optimized scopes, even when write
+  access wasn't needed or desired
+* that clarity of intent allows optimizing compilers and
+  interpreters to assume all local variable rebindings are
+  visible in the code, even when ``locals()`` is accessed
+  or the default arguments are used in APIs like ``exec()``
+  and ``eval()`` (optimization in this case refers to code
+  specialisation and other more advanced techniques, not
+  the basic optimization of replacing string dict lookups
+  with array index lookups - the latter feature is always
+  enabled in optimized scopes). Deoptimization will only
+  be needed when a write-through proxy instance is used to
+  modify a local variable. In a sufficiently sophisticated
+  interpreter, the deoptimization impact could potentially
+  be limited only to optimizations involving the modified
+  variable, but even less sophisticated interpreters will
+  be able to postpone frame deoptimization until at least
+  one local variable is modified outside of the normal flow
+  of code execution.
```

Comment:
I don't have any particular quibble with this bullet point, but to me it feels too far afield from this PEP to be speculating about possible compiler/interpreter optimizations. It just doesn't seem necessary to make the case.

Reply:
Keeping bindings visible in the code was definitely part of the rationale, and optimising compilers were one of the intended beneficiaries (human readers being the other).

I agree it feels out of place here, though. Maybe I can simplify the benefit of visible rebinding to something like "This improves clarity when reading code, and allows compilers and interpreters to more consistently apply additional performance optimizations in optimized scopes".

No need to get into the nitty gritty of interpreters implicitly reverting to less aggressive optimizations when hidden rebinding becomes a runtime concern.

Reply:
```suggestion
* ensuring name rebinding in optimized scopes remains visible in the code (as long as the frame introspection APIs are not accessed) improves clarity for human readers and allows compilers and interpreters to apply related performance optimizations more consistently.
```

Reply:
I changed this bullet point to the following:

```
* in addition to improving clarity for human readers, ensuring
  that name rebinding in optimized scopes remains lexically
  visible in the code (as long as the frame introspection APIs
  are not accessed) allows compilers and interpreters to apply
  related performance optimizations more consistently
```